File: app/src/main/java/io/neft/Renderer/Rectangle.java
package io.neft.Renderer;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;

import io.neft.Client.Action;
import io.neft.Client.InAction;
import io.neft.Client.Reader;
import io.neft.MainActivity;
import io.neft.Utils.ColorUtils;

public class Rectangle extends Item {
    static void register(final MainActivity app) {
        app.client.actions.put(InAction.CREATE_RECTANGLE, new Action() {
            @Override
            public void work(Reader reader) {
                new Rectangle(app);
            }
        });

        app.client.actions.put(InAction.SET_RECTANGLE_COLOR, new Action() {
            @Override
            public void work(Reader reader) {
                ((Rectangle) app.renderer.getItemFromReader(reader)).setColor(reader.getInteger());
            }
        });

        app.client.actions.put(InAction.SET_RECTANGLE_RADIUS, new Action() {
            @Override
            public void work(Reader reader) {
                ((Rectangle) app.renderer.getItemFromReader(reader)).setRadius(reader.getFloat());
            }
        });

        app.client.actions.put(InAction.SET_RECTANGLE_BORDER_COLOR, new Action() {
            @Override
            public void work(Reader reader) {
                ((Rectangle) app.renderer.getItemFromReader(reader)).setBorderColor(reader.getInteger());
            }
        });

        app.client.actions.put(InAction.SET_RECTANGLE_BORDER_WIDTH, new Action() {
            @Override
            public void work(Reader reader) {
                ((Rectangle) app.renderer.getItemFromReader(reader)).setBorderWidth(reader.getFloat());
            }
        });
    }

    static final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);

    public final Path path = new Path();
    public int color = Color.TRANSPARENT;
    public float radius = 0;
    public int borderColor = Color.TRANSPARENT;
    public float borderWidth = 0;

    public Rectangle(MainActivity app) {
        super(app);
    }

    protected void updatePath() {
        path.rewind();
        path.moveTo(radius, 0);
        path.lineTo(width - radius, 0);
        path.quadTo(width, 0, width, radius);
        path.lineTo(width, height - radius);
        path.quadTo(width, height, width - radius, height);
        path.lineTo(radius, height);
        path.quadTo(0, height, 0, height - radius);
        path.lineTo(0, radius);
        path.quadTo(0, 0, radius, 0);
    }

    @Override
    public void setWidth(float val) {
        super.setWidth(val);
        updatePath();
    }

    @Override
    public void setHeight(float val) {
        super.setHeight(val);
        updatePath();
    }

    public void setColor(int val) {
        color = ColorUtils.RGBAtoARGB(val);
        invalidate();
    }

    public void setRadius(float val) {
        radius = app.renderer.dpToPx(val);
        updatePath();
        invalidate();
    }

    public void setBorderColor(int val) {
        borderColor = ColorUtils.RGBAtoARGB(val);
        invalidate();
    }

    public void setBorderWidth(float val) {
        borderWidth = app.renderer.dpToPx(val);
        invalidate();
    }

    @Override
    protected void drawShape(final Canvas canvas, final int alpha) {
        paint.setAlpha(alpha);

        // fill
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(color);
        canvas.drawPath(path, paint);

        // stroke
        if (borderWidth > 0) {
            paint.setStyle(Paint.Style.STROKE);
            paint.setColor(borderColor);
            paint.setStrokeWidth(borderWidth);
            canvas.drawPath(path, paint);
        }
    }
}

File: Neft/Rectangle.swift
import SpriteKit

extension Renderer {
    class Rectangle: Renderer.BaseType {
        class Node: Item.Node {
            let shape = SKShapeNode()
            var radius: CGFloat = 0
            override var width: CGFloat {
                didSet {
                    self.updatePath()
                }
            }
            override var height: CGFloat {
                didSet {
                    self.updatePath()
                }
            }
            
            override init(_ renderer: Renderer) {
                super.init(renderer)
                self.shape.lineWidth = 0
                self.shape.lineJoin = .Miter
                self.childrenNode.addChild(self.shape)
            }

            required init?(coder aDecoder: NSCoder) {
                fatalError("init(coder:) has not been implemented")
            }
            
            private func updatePath(){
                let path = CGPathCreateMutable()
                let rect = CGRect(x: 0, y: -height, width: width, height: height)
                let radius = min(self.radius, width / 2, height / 2)
                if radius > 0 {
                    CGPathAddRoundedRect(path, nil, rect, radius, radius)
                } else {
                    CGPathAddRect(path, nil, rect)
                }
                shape.path = path
            }
        }
        
        override init(app: GameViewController){
            super.init(app: app)
            Renderer.actions[InActions.CREATE_RECTANGLE] = self.create
            Renderer.actions[InActions.SET_RECTANGLE_COLOR] = self.setColor
            Renderer.actions[InActions.SET_RECTANGLE_RADIUS] = self.setRadius
            Renderer.actions[InActions.SET_RECTANGLE_BORDER_COLOR] = self.setBorderColor
            Renderer.actions[InActions.SET_RECTANGLE_BORDER_WIDTH] = self.setBorderWidth
        }
        
        func create(reader: Reader){
            Node(renderer)
        }
        
        func setColor(reader: Reader){
            let item = renderer.getItem(reader) as! Node
            let color = reader.getInteger()
            item.shape.fillColor = hexColorToUIColor(color)
        }
        
        func setRadius(reader: Reader){
            let item = renderer.getItem(reader) as! Node
            let val = reader.getFloat()
            item.radius = val
            item.updatePath()
        }
        
        func setBorderColor(reader: Reader){
            let item = renderer.getItem(reader) as! Node
            let color = reader.getInteger()
            item.shape.strokeColor = hexColorToUIColor(color)
        }
        
        func setBorderWidth(reader: Reader){
            let item = renderer.getItem(reader) as! Node
            let val = reader.getFloat()
            item.shape.lineWidth = val
        }
    }
}
